packages/ai-chat-ui/src/browser/chat-input-widget.tsx

```diff
@@ -215,15 +222,40 @@ const ChatInput: React.FunctionComponent<ChatInputProperties> = (props: ChatInpu
         };
     }, []);
 
+    // track added requests
     React.useEffect(() => {
-        const listener = lastRequest?.response.onDidChange(() => {
-            if (lastRequest.response.isCanceled || lastRequest.response.isComplete || lastRequest.response.isError) {
-                setInProgress(false);
+        const onChatModelUpdate = (event: ChatChangeEvent) => {
+            if (event.kind === 'addRequest') {
+                onAddedRequest(() => event.request);
             }
-        });
+        };
+        const listener = props.chatModel.onDidChange(onChatModelUpdate);
+        return () => listener?.dispose();
+    }, []);
+
+    // remember last requests
+    React.useEffect(() => {
+        if (lastRequest) {
+            setInProgress(ChatRequestModel.isInProgress(lastRequest));
+        }
+        const onResponseUpdate = () => {
+            setInProgress(ChatRequestModel.isInProgress(lastRequest));
+        };
+        const listener = lastRequest?.response.onDidChange(onResponseUpdate);
         return () => listener?.dispose();
     }, [lastRequest]);
```

Comment:
This was not introduced by your PR but as I am just noticing it: We should never build `useEffect` based transformations for `state` which is under our control. At best this introduces unnecessary rerenderings and confusing state updates, at worst this ends up in endless rerender cycles.

As `lastRequest` is maintained by us, we should perform any side effects we want to see at the same time as `lastRrequest` is changed, i.e. by the code calling the setter.